File: 0101-symmetric-tree/0101-symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    bool check(TreeNode* node1, TreeNode* node2) {
        if ((node1 == nullptr && node2 != nullptr) ||
            (node2 == nullptr && node1 != nullptr))
            return false;
        else if (node1 == nullptr && node2 == nullptr)
            return true;
        else if (node1->val != node2->val)
            return false;
        else {
            bool outer = check(node1->left, node2->right);
            bool inner = check(node1->right, node2->left);
            return outer && inner;
        }
    }

    bool isSymmetric(TreeNode* root) {
        if (root == nullptr)
            return true;
        return check(root->left, root->right);
    }
};

File: 0101-symmetric-tree/0101-symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        # Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def check(self, node1, node2):
        if node1 is None and node2 is None:
            return True
        elif node1 is None or node2  is None:
            return False
        elif node1.val != node2.val:
            return False
        else:
            return self.check(node1.left, node2.right) and self.check(node1.right, node2.left)

    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None:
            return True
        return self.check(root.left, root.right)
        